dist/sameorigin_links.js

```diff
@@ -0,0 +1,71 @@
+const intersectRect = function (r1, r2) {
+    return !(r2.left > r1.right || 
+              r2.right < r1.left || 
+              r2.top > r1.bottom ||
+              r2.bottom < r1.top);
+};
+const viewport = {
+    left: 0,
+    top: 0,
+    right: (window.innerWidth || document.documentElement.clientWidth),
+    bottom: (window.innerHeight || document.documentElement.clientHeight)
+};
+const getBoundingClientRect = function(element) { 
+    const rect = element.getBoundingClientRect()
+    return {
+        top: parseInt(rect.top),
+        right: parseInt(rect.right),
+        bottom: parseInt(rect.bottom),
+        left: parseInt(rect.left),
+        width: parseInt(rect.width),
+        height: parseInt(rect.height),
+        x: parseInt(rect.x),
+        y: parseInt(rect.y)
+      };
+};
+const sameOrigin = function(uri1, uri2){
+    uri1 = new URL(uri1);
+    uri2 = new URL(uri2);
+    if(uri1.host !== uri2.host) return false;
+    if(uri1.port !== uri2.port) return false;
+    if(uri1.protocol !== uri2.protocol) return false;
+    return true;
+}
+let links = {};
+const elementTypes = ['img', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6'];
+const documentOrigin = window.location.href;
+for (let eType of elementTypes) {
```

Comment:
This outer loop could be replaced with a more complex querySelectorAll:

```js
let elements = document.querySelectorAll('img, h1, h2, h3, h4, h5, h6');
```

Reply:
It could, but I'm storing them under different indexes and sorting each by size so it was cleaner to keep it as a separate loop, otherwise the sorting will need its own loop and we'll need to extract the element type from each result.